app/src/main/java/unipd/se18/ocrcamera/inci/NameMatchIngredientsExtractor.java

```diff
@@ -0,0 +1,96 @@
+package unipd.se18.ocrcamera.inci;
+
+import android.util.Log;
+
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+
+
+/**
+ * IngredientsExtractor implementation that for each INCI ingredient check if it is contained
+ * inside the text. Non alphanumeric characters are ignored.
+ * @author Francesco Pham
+ */
+public class NameMatchIngredientsExtractor implements IngredientsExtractor {
+
+    //list where all ingredients from inci db are stored
+    private List<Ingredient> listIngredients;
+
+    //Tag for logs
+    private static final String TAG = "IngredientsExtractor";
+
+    /**
+     * Constructor initializes the ingredients extractor
+     * @param listIngredients Total list of ingredients from the INCI DB
+     */
+    public NameMatchIngredientsExtractor(List<Ingredient> listIngredients) {
+        //copying list so that sorting doesn't affect original list
+        this.listIngredients = new ArrayList<>(listIngredients);
+
+        //sort by name length so when we search for ingredients in text we match longer names first
+        Collections.sort(this.listIngredients, new Comparator<Ingredient>() {
+            @Override
+            public int compare(Ingredient o1, Ingredient o2) {
+                return o2.getInciName().length() - o1.getInciName().length();
+            }
+        });
+    }
+
+
+    /**
+     * This method extracts ingredients from the ocr text and returns the list of ingredients.
+     * @param text The entire OCR text
+     * @return List of extracted ingredients, empty list if no ingredients are found
+     * @author Francesco Pham
+     */
+    @Override
+    public List<Ingredient> findListIngredients(String text) {
+
+        List<Ingredient> foundIngredients = new ArrayList<>();
+
+        //remove non alphanumeric characters from text
+        //in mapIndexes we store for each character in the stripped text, the original position
+        int[] mapIndexes = new int[text.length()];
+        StringBuilder strippedTextBuilder = new StringBuilder();
+        for(int i=0; i<text.length(); i++) {
+            char currentChar = text.charAt(i);
+            if(Character.isLetter(currentChar) || Character.isDigit(currentChar)) {
+                mapIndexes[strippedTextBuilder.length()] = i;
+                strippedTextBuilder.append(currentChar);
+            }
+        }
```

Comment:
Why dont u just use String.replace('\W','') instead of map it and use this complex method.
Anyway if it's necessary could be better to create a new method that do that to make the code more linear and to make the code more modular

Reply:
Because I need to map the indexes between the original text and the text without symbols (because later I need the position of the ingredient name in the original text). If I did `String.replace('\W','')` I would loose any of these informations. 
I didn't think it was much code to create a seperate method but if you think it would be more clearer than I will do it, keep in mind that this new method would need to return two things that are needed in the rest of the code: mapIndexes and the strippedText. So If I had to make a separate method then I would have to add a class variable.